lib/web_ui/lib/src/engine/skwasm/skwasm_impl/renderer.dart

```diff
@@ -285,9 +295,85 @@ class SkwasmRenderer implements Renderer {
       indices: indices
     );
 
+  ui.Size? _scaledSize(
+    int width,
+    int height,
+    int? targetWidth,
+    int? targetHeight,
+  ) {
+    if (targetWidth == null && targetHeight == null) {
+      return null;
+    }
+    if (targetWidth == width && targetHeight == height) {
+      // Not scaled
+      return null;
+    }
+    if (targetWidth == null) {
+      if (targetHeight == height) {
+        // Not scaled.
+        return null;
+      }
+      targetWidth = (width * targetHeight! / height).round();
+    } else if (targetHeight == null) {
+      if (targetWidth == targetWidth) {
+        // Not scaled.
+        return null;
+      }
+      targetHeight = (height * targetWidth / width).round();
+    }
+    return ui.Size(targetWidth.toDouble(), targetHeight.toDouble());
+  }
+
   @override
-  void decodeImageFromPixels(Uint8List pixels, int width, int height, ui.PixelFormat format, ui.ImageDecoderCallback callback, {int? rowBytes, int? targetWidth, int? targetHeight, bool allowUpscaling = true}) {
-    throw UnimplementedError('decodeImageFromPixels not yet implemented');
+  void decodeImageFromPixels(
+    Uint8List pixels,
+    int width,
+    int height,
+    ui.PixelFormat format,
+    ui.ImageDecoderCallback callback, {
+    int? rowBytes,
+    int? targetWidth,
+    int? targetHeight,
+    bool allowUpscaling = true
+  }) {
+    final ui.Size? scaledSize = _scaledSize(
+      width,
+      height,
+      targetWidth,
+      targetHeight
+    );
+    if (!allowUpscaling && scaledSize != null &&
+      (scaledSize.width > width || scaledSize.height > height)) {
+        domWindow.console.warn('Cannot apply targetWidth/targetHeight when allowUpscaling is false.');
```

Comment:
I'm not sure we need to issue a warning. This seems to be the expected behavior. Sometimes you want to shrink the image to save memory, but upscaling can be done in the GPU so you don't want to waste CPU and memory on upscaling prior to rendering.

Reply:
I was doing this to be consistent with `skiaDecodeImageFromPixels`, which issues this warning under this condition.

Reply:
Heh, benefits of revisiting old code! :) Yeah, I don't think `skiaDecodeImageFromPixels` should warn either.